full/Dockerfile

```diff
@@ -1,192 +1,153 @@
+        php-ctype \
+        php-curl \
+        php-date \
         php-gd \
+        php-gettext \
+        php-intl \
         php-json \
-        php-curl \
         php-mbstring \
+        php-mysql \
+        php-net-ftp \
+        php-pgsql \
         php-sqlite3 \
         php-tokenizer \
-        php-ctype \
-        php-date \
-        php-net-ftp \
-        php-gettext \
+        php-xml \
         php-zip \
-    && apt-get clean && rm -rf /var/cache/apt/* && rm -rf /var/lib/apt/lists/* && rm -rf /tmp/*
- # the PHP language server is installed by the theia-php-extension
+    && apt-get clean && rm -rf /var/lib/apt/lists/* /tmp/*
+# PHP language server is installed by theia-php-extension
+
+### Yarn ###
+RUN curl -fsSL https://dl.yarnpkg.com/debian/pubkey.gpg | apt-key add - \
+    && apt-add-repository -yu "deb https://dl.yarnpkg.com/debian/ stable main" \
+    && apt-get install --no-install-recommends -yq yarn=1.12.3-1 \
+    && apt-get clean && rm -rf /var/lib/apt/lists/* /tmp/*
+
+### Gitpod user (2) ###
+USER gitpod
+# use sudo so that user does not get sudo usage info on (the first) login
+RUN sudo echo "Running 'sudo' for Gitpod: success"
+
+### Go ###
+ENV GO_VERSION=1.11.2 \
+    GOPATH=$HOME/go-packages \
+    GOROOT=$HOME/go
+ENV PATH=$GOROOT/bin:$GOPATH/bin:$PATH
+RUN curl -fsSL https://storage.googleapis.com/golang/go$GO_VERSION.linux-amd64.tar.gz | tar -xzv \
+    && go get -u -v \
+        github.com/acroca/go-symbols \
+        github.com/cweill/gotests/... \
+        github.com/davidrjenni/reftools/cmd/fillstruct \
+        github.com/fatih/gomodifytags \
+        github.com/haya14busa/goplay/cmd/goplay \
+        github.com/josharian/impl \
+        github.com/nsf/gocode \
+        github.com/ramya-rao-a/go-outline \
+        github.com/rogpeppe/godef \
+        github.com/uudashr/gopkgs/cmd/gopkgs \
+        github.com/zmb3/gogetdoc \
+        golang.org/x/lint/golint \
+        golang.org/x/tools/cmd/godoc \
+        golang.org/x/tools/cmd/gorename \
+        golang.org/x/tools/cmd/guru \
+        sourcegraph.com/sqs/goreturns
+# user Go packages
+ENV GOPATH=/workspace:$GOPATH \
+    PATH=/workspace/bin:$PATH
+
+### Node.js ###
+ARG NODE_VERSION=8.14.0
+ENV PATH=/home/gitpod/.nvm/versions/node/v8.14.0/bin:$PATH
+RUN curl -fsSL https://raw.githubusercontent.com/creationix/nvm/v0.33.11/install.sh | bash \
+    && bash -c ". .nvm/nvm.sh \
+        && npm config set python /usr/bin/python --global \
+        && npm config set python /usr/bin/python \
+        && npm install -g typescript"
+
+### Python ###
+ENV PYENV_ROOT=$HOME/.pyenv
+ENV PATH=$PYENV_ROOT/bin:$PATH
+RUN curl -fsSL https://github.com/pyenv/pyenv-installer/raw/master/bin/pyenv-installer | bash \
+    && { echo; echo 'eval "$(pyenv init -)"'; echo 'eval "$(pyenv virtualenv-init -)"'; } >> .bashrc \
+    && pyenv install 3.6.6 \
+    && pyenv global 3.6.6 \
+    && $PYENV_ROOT/shims/pip install virtualenv python-language-server[all]==0.19.0 \
```

Comment:
I think it's not good to call a shim directly.

after `pyenv global 3.6.6` a call to `pip` is using it automatically, and you shouldn't care about that.

also, if you remove it from PATH, how should that be found then in general?

Reply:
I'll double-check.

Reply:
I guess that's the reason for having an explicit ENV statement for pyenv.

Reply:
I checked the `pyenv` installation steps and it's like that:
  1. you install pyenv and manually modify `~/.bashrc`
  2. you restart the shell; `pyenv`, all shims etc. will now be in PATH as that is modified by the newly added lines to '~/.bashrc'
  3. you do the remaining part of what we want to install (Python 3.6.6, `virtual-env`)

But for us, even if we split that single `RUN` in two, I found no nice way to start `bash` so that it will source `.bashrc`, so basically we can never get that dynamically made changes work for us. Therefore I call the two executables (`pyenv` and the shim `pip`) directly in the Dockerfile. (Before we put both directories in the `PATH` explicitly, so in the end for both the dirs the `$PATH` was duplicated at runtime - that most probably wouldn't mind anything, but it's ugly at least)

Reply:
cannot you just run `source ~/.bashrc && pip install`?